### PR TITLE
Flowers: Fix waterlily node's duplication with stack of 1

### PR DESCRIPTION
**Pull request by kilbith**
Fixes https://github.com/minetest/minetest_game/issues/1046.

Cc: @paramat

### CODE REVIEW

**Comment by PilzAdam** [comment]
Why is this necessary? What is the underlying problem for this hack?

**Comment by kilbith** [comment]
The answer is somewhere in the lines of C++, but anyway returning the `itemstack` triggers the bug.

**Comment by paramat** [comment]
Thanks, much appreciated, will test.

**Comment by Zeno-** [comment]
(referring to bug report as well because the comments in it are relevant)

If this is a workaround for release only, how will we remember to look for the true cause after release?

**Comment by HybridDog** [comment]
screenshot_20160422_100305
l didn't manage to duplicate the waterlily, l had one left in survival mode, placed it and then it was taken from inventory.
There was just a bug with node placement prediction, so it looks like another waterlily is under the placed one, after rejoining the other one is gone.
Maybe you only need to execute nodeupdate() if the last waterlily gets placed.

File: mods/flowers/init.lua [hunk 1]
```diff
@@ -235,15 +235,22 @@ minetest.register_node("flowers:waterlily", {
...
 		local def = minetest.registered_nodes[node]
 		if def and def.liquidtype == "source" and minetest.get_item_group(node, "water") > 0 then
 			minetest.set_node(pos, {name = "flowers:waterlily", param2 = math.random(0, 3)})
 			if not minetest.setting_getbool("creative_mode") then
-				itemstack:take_item()
-				return itemstack
+				if itemstack:get_count() == 1 then
+					local inv = placer:get_inventory()
+					minetest.after(0, function()
+						inv:set_stack("main", placer:get_wield_index(), nil)
```
Comment (HybridDog) [inline mods/flowers/init.lua]: doing it this way is dangerous, if the player places it and then switches to another item, the wrong one gets removed

**Comment by HybridDog** [comment]
How does placing waterlilies look like if the server lags?

File: mods/flowers/init.lua (hunk 1 above)
Comment (kilbith) [inline mods/flowers/init.lua]: Indeed, it's better to localize before.

**Comment by paramat** [comment]
> If this is a workaround for release only, how will we remember to look for the true cause after release?

We can keep the issue open when this is merged, i've added a comment in the issue.

**Comment by paramat** [comment]
Tested :+1:

**Comment by Ekdohibs** [comment]
The code is still dangerous, if the player inventory changes right after the player places the lilypad, then the wrong item will be removed. Isn't there a way to disable lily pad placement prediction instead? (Which is wrong anyway).

**Comment by paramat** [comment]
Ekdohibs proposal sounds better to me, it would also stop the waterlily appearing underwater briefly.

**Comment by kilbith** [comment]
Applied @Ekdohibs suggestion, which is much better.

**Comment by Ekdohibs** [comment]
:+1: from me now

**Comment by paramat** [comment]
Testing.

**Comment by paramat** [comment]
Seems good :+1:

File: mods/flowers/init.lua [hunk 2]
```diff
@@ -235,7 +236,7 @@ minetest.register_node("flowers:waterlily", {
 		fixed = {-0.5, -0.5, -0.5, 0.5, -0.4375, 0.5}
 	},
 
-	on_place = function(itemstack, _, pointed_thing)
+	on_place = function(itemstack, placer, pointed_thing)
```
Comment (HybridDog) [inline mods/flowers/init.lua]: the player could be used for testing if there's protection

**Comment by kilbith** [comment]
Protection check added.

File: mods/flowers/init.lua (hunk 2 above)
Comment (kilbith) [inline mods/flowers/init.lua]: Thanks for deleting this comment.

**Comment by paramat** [comment]
:+1:

**Comment by paramat** [comment]
16fe1efcd82a8274989962bf19f7d6f7befd3cfe

**Pull request closed without merging**